### PR TITLE
fix: add promise queues for providers

### LINKED ISSUE #2613: [BUG] Sometimes window.nostr/liquid/webln takes too long to load and is unavailable when website checks for it
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

The inpage script is sometimes loaded too slowly because it is loaded asynchronously, causing sites to think the objects aren't available.

### Screenshots [optional]

_No response_

### Steps To Reproduce

1. Visit https://louisinger.github.io/alby-liquid-wallet/#
2. Sometimes it will throw an exception because window.liquid does not exist

### Expected behavior

window.liquid/nostr/webln should be available instantly.

### Alby information

Latest

### Device information

_No response_

### Additional context

https://github.com/getAlby/lightning-browser-extension/issues/2462

### Are you working on this?

None

### PR DESCRIPTION
**Pull request by reneaaron**
### Describe the changes you have made in this PR

Fixes #2613

### TODO
 - Implement queue for other providers

### CODE REVIEW

File: src/extension/providers/nostr/index.ts [hunk 1]
```diff
@@ -1,21 +1,33 @@
...
 import { Event } from "./types";
-import EventEmitter from "events";
 
 declare global {
   interface Window {
     nostr: NostrProvider;
   }
 }
 
+class PromiseQueue {
```
Comment (rolznz) [inline src/extension/providers/nostr/index.ts:11]: Can you move this class to a new file?

File: src/extension/providers/nostr/index.ts [hunk 2]
```diff
@@ -1,21 +1,33 @@
...
+      this.queue = this.queue.then(operation).then(resolve).catch(reject);
+    });
+  }
+}
+
 export default class NostrProvider {
   nip04 = new Nip04(this);
   enabled: boolean;
   private _eventEmitter: EventEmitter;
+  queue: PromiseQueue;
```
Comment (rolznz) [inline src/extension/providers/nostr/index.ts:25]: this should be private

File: src/extension/providers/nostr/index.ts [hunk 3]
```diff
@@ -30,23 +42,31 @@ export default class NostrProvider {
...
+    });
   }
 
   async getRelays() {
-    await this.enable();
-    return this.execute("getRelays");
+    return this.queue.add(async () => {
+      await this.enable();
+      return this.execute("getRelays");
+    });
```
Comment (rolznz) [inline src/extension/providers/nostr/index.ts:69]: `enable` in the on and off methods should also be wrapped in the queue, right?

Comment (rolznz) [reply src/extension/providers/nostr/index.ts:69]: For this reason I am wondering if the execute method itself should handle the queue, hmm..

Comment (bumi) [reply src/extension/providers/nostr/index.ts:69]: I was wondering the same. we always call execute, could execute implement what is currently in the `PromiseQueue #add`

Comment (reneaaron) [reply src/extension/providers/nostr/index.ts:69]: Yes, but maybe we could even move it to execute<T>(), I'll check.

**Review by bumi — APPROVED** [review]

**Pull request merged**